[ready] overhaul -j to make it more flexible, staying 100% compatible

mirabilos opened the pull request:
## **Update:** This is ready to be merged.

[…] This will also make the `-j` option usable much better and, for the most part, eliminate the need to use `-P` with some of the scores and not use it with the others (see [the doc forum post about `-j`](https://musescore.org/en/node/276949)).

Basically, I can just generate a JSON that lists *all* to-be-converted scores, in both the string and two-element-array function, and the first will generate the main score PDF, and the latter will generate parts PDFs if the score has parts and otherwise be silently ignored. It is also more flexible than adding `-partname` before the file extension as `exportParts()` does. (Besides, `exportParts()` was not exposed to the command line *at all*, and I believe I found a great way to do so.)

Here’s now the commit message:

    overhaul -j to make it more flexible, staying 100% compatible
    
    The previous operation of
    [ { "in": "file.mscz", "out": "file.pdf", "plugin": "foo.qml" } ]
    is still supported, and nothing changes: either “out” or “plugin”
    must be given, and -P appends all parts to PDF and PNG, generating
    parts for ALL instruments if the score has none.
    
    Because people often write this:
    [ { "in": "a.mscz", "out": "a.pdf" },
      { "in": "a.mscz", "out": "a.mid" },
      { "in": "b.mscz", "out": "b.pdf" },
      { "in": "b.mscz", "out": "b.mid" } ]
    
    We now also support shortening it like this:
    [ { "in": "a.mscz", "out": [ "a.pdf", "a.mid" ] },
      { "in": "b.mscz", "out": [ "b.pdf", "b.mid" ] } ]
    
    This new syntax saves the same score twice, i.e. doesn’t reload
    nor re-apply the plugin in between. (We assume that saving does
    not modify the score.)
    
    Finally, there’s a new syntax intended to be used WITHOUT -P:
    [ { "in": "file.mscz", "out": [ [ "file (part for ", ").pdf" ] ] } ]
    
    Explanation: iff “out” is an array (new syntax), and one of its
    elements is again an array, with two string members, then all
    extant parts are exported with the part name inserted between
    the two strings (and if the score has no parts, the entry is
    silently ignored, parts are NOT auto-generated here). This works
    with all file types, not just PNG or PDF, and does NOT export
    the master score, ONLY the parts.
    
    (I also intend to document this in the manual in more detail.)
    
    Intended use is like this:
    [ { "in": "file.mscz", "out": [
            "file.pdf",
            "file.mid",
            "file.musicxml",
            [ "file (part for ", ").pdf" ],
            [ "file (part for ", ").mid" ]
      ] } ]
    
    All these syntacēs can be mixed in the same job file (although
    only if the job is run without the -P option):
    
    [
      {
        "in": "demos/Triumph.mscz",
        "out": "Triumph-coloured.pdf",
        "plugin": "colornotes.qml"
      },
      {
        "in": "demos/Reunion.mscz",
        "out": [
          "Reunion.pdf",
          "Reunion.musicxml",
          "Reunion.mid"
        ]
      },
      {
        "in": "piece.mscz",
        "out": [
          "piece.pdf",
          "piece.mid",
          "piece.musicxml",
          [
            "piece (part for ",
            ").pdf"
          ],
          [
            "piece (part for ",
            ").mid"
          ]
        ]
      }
    ]

Jojo-Schmitz commented on mscore/musescore.cpp:
```diff
@@ -3210,39 +3213,79 @@ static bool doConvert(Score* cs, QString fn, QString plugin = "")
                   cs->doLayout();
                   }
             }
+      foreach (const QJsonValue& outFile, outFiles) {
```
better use `for (const QJsonValue& outFile : outFiles) {`

mirabilos replied:
Thanks, fixed in the next push.

Jojo-Schmitz commented on mscore/musescore.cpp:
```diff
@@ -3210,39 +3213,79 @@ static bool doConvert(Score* cs, QString fn, QString plugin = "")
+                  // convert parts
+                  QString fnbeg = fns[0].toString();
+                  QString fnend = fns[1].toString();
+                  foreach (Excerpt* e, cs->excerpts()) {
```
similar here

mirabilos commented:
@Jojo-Schmitz I’m currently testing a backport to 2.3 and found redundant code. Do you think https://github.com/musescore/MuseScore/pull/4166/commits/cd40fca24cec82d635ba1f24fc7118d2560375ff should be added to this (as it touches the same area)?

Rationale: we already set the page layout at the beginning of working on the score and (if needed) do a full relayout. (Also, don’t do the same thing twice in a row.) The layout is reset at the end of the function already anyway, the if in the plugin block shouldn’t even trigger.

Also, did you have any luck testing this on master? I got a segfault during early tests with the 2.3 backport (but probably more related to 2.3 than to this diff), so testing is crucial. I’m testing with:

    [
      {
        "in": "mbsd/mirsol/music/free/Kuhlo -- Ein feste Burg.mscx",
        "out": "1-normal.pdf"
      },
      {
        "in": "mbsd/mirsol/music/free/Kuhlo -- Ein feste Burg.mscx",
        "out": "1-plugin.pdf",
        "plugin": "colornotes.qml"
      },
      {
        "in": "/usr/share/mscore-2.3/demos/Reunion.mscz",
        "out": [
          "Reunion.pdf",
          "Reunion.mid"
        ]
      },
      {
        "in": "mbsd/mirsol/music/chor/Michel -- Macht hoch die Tür + Hosianna.mscx",
        "out": [
          "piece.pdf",
          "piece.mid",
          "piece.musicxml",
          [
            "piece (part for ",
            ").pdf"
          ],
          [
            "piece (part for ",
            ").mid"
          ]
        ]
      },
      {
        "in": "mbsd/mirsol/music/free/Silcher -- Loreley.mscx",
        "out": [
          "nopiece.pdf",
          "nopiece.mid",
          "nopiece.musicxml",
          [
            "nopiece (part for ",
            ").pdf"
          ],
          [
            "nopiece (part for ",
            ").mid"
          ]
        ]
      }
    ]

Here I have a score with and one without excerpts in the bottom.

I’ll resolve the branch conflict later (I know why, @lasconic added a file type, but that’s fine and easy enough to handle).

Jojo-Schmitz commented:
I'd change 2.3 as little as possible, if at all

mirabilos commented:
I now successfully tested a backport of this to 2.3 and even Valgrind is now happy and does not show any memory access violations any more.

Some testing of this on master would still be welcome. The commit message has an example input JSON file. It will take me a while if I have to do it myself.

Jojo-Schmitz commented:
how about adding an mtest for it?

Jojo-Schmitz commented:
Maybe the `share/templates/convert.json` from  #4107 could help as a test case?

mirabilos commented:
It’d be a start. We also need an example with excerpts, to test part extraction. The results need manual validation though (cursory look is enough, just so that the colournotes plugin was run on one of them, and that the PDF and MIDI and MusicXML files’ content is the expected), in addition to just “not segfaulting” and “the names of the resulting files are correct” (especially in the case where part extraction is silently ignored for the example that doesn’t have parts). It’s probably too much effort to be worth it, and it really has to be tested only once, for now.

I’m under $dayjob deadline for multiple projects and will likely spend the entire weekend at work, though.

mirabilos commented:
OK, I figured out things, and the crashes when exporting MIDIs for parts is independent of this PR (but a separate bug), so this PR is now good to go.

Eventually (when they will be fixed) we’ll wish to re-enable deleting the just-processed score, but I watched memory consumption batch-converting 55 scores to about 250 destination files, and, even with the added Valgrind overhead, it’s very manageable.

mirabilos commented:
Updated to close/`delete` the score again after conversation of it has finished. This was made possible by @AntonioBL and @dmitrio95 their PRs which fix all the memory corruption.

Tested in Valgrind with this patch applied, converted 52 scores (1 `.mscz`, 51 `.mscx`) in 2.x format to 51 `.mscz`, 52 `.mid`, 52 `.xml` (MusicXML), and 54 `.pdf` (52 score, 2 extracts/parts). I do not have any 3.x format test cases, although I *did* test this on the samples (Reunion and Triumph… oh, these are also 2.x scores).

So, please apply ;-)

mirabilos commented:
Rebased on git master as of today (3.0.1 release equivalent), please apply now.

mirabilos commented:
@anatoly-os ping?

anatoly-os commented:
I would really appreciate some tests which support such refactoring. Did you test it manually? What scenarios did you use?

mirabilos commented:
I tested it with the JSON I gave in the longer description and several (2.x though) scores. The output matches what is expected.

mirabilos commented:
… oh, and I tested a larger export on several dozen scores (all I have), with PDF, MusicXML, MIDI export for the scores and PDF export for all existing parts. I didn’t test with plugins though, as they were not working at that time, but I *did* test this excessively, also with plugins, in 2.3.2.
